sdks/go/pkg/beam/registration.tmpl

```diff
@@ -0,0 +1,292 @@
+
+        setupWrapper = func(fn interface{}) reflectx.Func {
+            return reflectx.MakeFunc(func(a0 context.Context) error {
+                return fn.(setup1x1).Setup(a0)
+            })
+        }
+    }
+    var teardownWrapper func(fn interface{}) reflectx.Func
+    if _, ok := doFn.(teardown0x0); ok {
+        teardownCaller := func(fn interface{}) reflectx.Func {
+            f := fn.(func())
+            return &caller0x0{fn: f}
+        }
+        reflectx.RegisterFunc(reflect.TypeOf((*func())(nil)).Elem(), teardownCaller)
+
+        teardownWrapper = func(fn interface{}) reflectx.Func {
+            return reflectx.MakeFunc(func() {
+                fn.(teardown0x0).Teardown()
+            })
+        }
+    } else if _, ok := doFn.(teardown1x0); ok {
+        teardownCaller := func(fn interface{}) reflectx.Func {
+            f := fn.(func(context.Context))
+            return &caller1x0[context.Context]{fn: f}
+        }
+        reflectx.RegisterFunc(reflect.TypeOf((*func(context.Context))(nil)).Elem(), teardownCaller)
+
+        teardownWrapper = func(fn interface{}) reflectx.Func {
+            return reflectx.MakeFunc(func(a0 context.Context) {
+                fn.(teardown1x0).Teardown(a0)
+            })
+        }
+    } else if _, ok := doFn.(teardown0x1); ok {
+        teardownCaller := func(fn interface{}) reflectx.Func {
+            f := fn.(func() error)
+            return &caller0x1[error]{fn: f}
+        }
+        reflectx.RegisterFunc(reflect.TypeOf((*func() error)(nil)).Elem(), teardownCaller)
+
+        teardownWrapper = func(fn interface{}) reflectx.Func {
+            return reflectx.MakeFunc(func() error {
+                return fn.(teardown0x1).Teardown()
+            })
+        }
+    } else if _, ok := doFn.(teardown1x1); ok {
+        teardownCaller := func(fn interface{}) reflectx.Func {
+            f := fn.(func(context.Context) error)
+            return &caller1x1[context.Context, error]{fn: f}
+        }
+        reflectx.RegisterFunc(reflect.TypeOf((*func(context.Context) error)(nil)).Elem(), teardownCaller)
+
+        teardownWrapper = func(fn interface{}) reflectx.Func {
+            return reflectx.MakeFunc(func(a0 context.Context) error {
+                return fn.(teardown1x1).Teardown(a0)
+            })
+        }
+    }
+    wrapperFn := func(fn interface{}) map[string]reflectx.Func {
+        m := map[string]reflectx.Func{}
+        if processElementWrapper != nil {
+            m["ProcessElement"] = processElementWrapper(fn)
+        }
+        if startBundleWrapper != nil {
+            m["StartBundle"] = startBundleWrapper(fn)
+        }
+        if finishBundleWrapper != nil {
+            m["FinishBundle"] = finishBundleWrapper(fn)
+        }
+        if setupWrapper != nil {
+            m["Setup"] = setupWrapper(fn)
+        }
+        if teardownWrapper != nil {
+            m["Teardown"] = teardownWrapper(fn)
+        }
+        
+        return m
+    }
```

Comment:
This is probably something you had in mind for helping reduce the size of the generated code since a quick skim reads pretty static...

Do recall that if a helper doesn't need arity assistance, it may be able to be put into a non-generated file the same package, and tested normally.

Reply:
This was not actually one the big one that I was considering, though I did change it as well

> Do recall that if a helper doesn't need arity assistance, it may be able to be put into a non-generated file the same package, and tested normally.

That brings up an interesting style question - I had considered doing this initially, but decided to keep it in the template file because logically it belongs with the other functions in registration.go. Stylistically, is it more typical to move functions like this out of the template file or to keep them with the functions they logically go with? Doesn't really matter for this case at this point, but I'm curious about the principle